user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/components/experimentalvariable/ExperimentalVariablesInput.java

```diff
@@ -0,0 +1,324 @@
+        .filter(VariableRow.class::isInstance)
+        .map(VariableRow.class::cast)
+        .toList();
+  }
+
+  public void addVariable() {
+    VariableRow variableRow = VariableRow.empty();
+    addRow(variableRow);
+  }
+
+  public void addVariable(ExperimentalVariableInformation variableInformation) {
+    VariableRow variableRow = VariableRow.filled(variableInformation);
+    addRow(variableRow);
+  }
+
+  private void addRow(VariableRow variableRow) {
+    variableRow.addDeleteVariableListener(it -> {
+      if (it.getSource().isEmpty()) {
+        removeVariable(it.getSource());
+        return;
+      }
+      AppDialog confirmDialog = confirmVariableDeletionDialog(it.getSource().getVariableName());
+      confirmDialog.registerConfirmAction(() -> {
+        removeVariable(it.getSource());
+        confirmDialog.close();
+      });
+      confirmDialog.registerCancelAction(confirmDialog::close);
+      confirmDialog.open();
+    });
+    var maxRowIndex = variablesInput.getChildren()
+        .filter(VariableRow.class::isInstance)
+        .mapToInt(variablesInput::indexOf)
+        .max().orElse(-1);
+    variablesInput.addComponentAtIndex(Math.max(0, maxRowIndex + 1), variableRow);
+    variableRow.focus();
+  }
+
+  private static AppDialog confirmVariableDeletionDialog(String variableName) {
+    AppDialog confirmDialog = AppDialog.small();
+    DialogHeader.withIcon(confirmDialog, "Delete experimental variable?",
+        IconFactory.warningIcon());
+    Span variableNameText = new Span(variableName);
+    variableNameText.addClassNames("bold padding-horizontal-02");
+    DialogBody.withoutUserInput(confirmDialog,
+        new Div(new Span("The variable"), variableNameText,
+            new Span("will be deleted.")));
+    DialogFooter.withDangerousConfirm(confirmDialog, "Cancel",
+        "Delete Experimental Variable");
+    return confirmDialog;
+  }
+
+  private void removeVariable(@NonNull VariableRow variableRow) {
+    if (!variableRow.isEmpty()) {
+      var initialComponentState = toRestoredVariablesInput(this.initialState);
+      var initialRowInformation = toRestoredVariableRow(variableRow.getInitialState());
+      if (initialComponentState.getVariableRows().stream()
+          .anyMatch(row -> row.getVariableName().equals(initialRowInformation.getVariableName()))) {
+        //only record deletion of variable rows that were there initially
+        var snapshotOfVariableRow = variableRow.snapshot();
+        deletedVariables.add(snapshotOfVariableRow);
+      }
+    }
+    variableRow.removeFromParent();
+    if (getVariableRows().isEmpty()) {
+      addVariable();
+    }
+  }
+
+  private ExperimentalVariablesInput toRestoredVariablesInput(Snapshot snapshot) {
+    try {
+      var dummy = new ExperimentalVariablesInput();
+      dummy.restore(snapshot);
+      return dummy;
+    } catch (SnapshotRestorationException e) {
+      throw new IllegalStateException("Initial state not correctly set.", e);
```

Comment:
it is not an illegal state. It is an illegal argument at most or let the snapshot exception bubble up to the caller, which is the same class. Otherwise i dont see the point of having this semantic exception at all.

Reply:
addressed